src/toolkits/drawing_classifier/dc_data_iterator.hpp

```diff
@@ -0,0 +1,164 @@
+     */
+    flex_list class_labels;
+
+    /**  Set to true, when the data is used for training. */
+    bool is_train = false;
+
+    /** Determines results of data augmentation if enabled. */
+    int random_seed = 0;
+  };
+
+  /** Defines the output of a data_iterator. */
+  struct batch {
+
+    /** Defines the metadata associated with each chunk. */
+    struct chunk_info {
+
+      /** Number of samples (rows from the raw SFrame) comprising the chunk. */
+      size_t num_samples;
+    };
+
+    /**
+     * TODO: Add explanatory comment 
+     */
+    neural_net::shared_float_array features;
+
+    /**
+     * TODO: Add explanatory comment 
+     */
+    neural_net::shared_float_array labels;
+
+    /**
+     * TODO: Add explanatory comment 
+     */
+    neural_net::shared_float_array weights;
+
+
+    /**
+     * TODO: Add explanatory comment 
+     */
+    neural_net::shared_float_array labels_per_row;
+
+    /**
+     * TODO: Add explanatory comment 
+     */
+    std::vector<chunk_info> batch_info;
+
+  };
+
+  virtual ~data_iterator();
+
+  virtual const flex_list& feature_names() const = 0;
+  virtual const flex_list& class_labels() const = 0;
+  virtual flex_type_enum session_id_type() const = 0;
+
+  /**
+   * Returns true if and only if the next call to `next_batch` will return a
+   * batch with size greater than 0.
+   */
+  virtual bool has_next_batch() const = 0;
+
+  /**
+   * Returns a batch containing float arrays with the indicated batch size.
+   *
+   * Eventually returns a batch with size smaller than the requested size,
+   * indicating that the entire dataset has been traversed. All subsequent
+   * calls will return an empty (all padding) batch, until reset.
+   */
+  virtual batch next_batch(size_t batch_size) = 0;
+
+  /** Begins a fresh traversal of the dataset. */
+  virtual void reset() = 0;
+};
+
+/**
+ * Concrete data_iterator implementation that doesn't attempt any
+ * parallelization or background I/O.
+ */
+class simple_data_iterator: public data_iterator {
+public:
+
+  simple_data_iterator(const parameters& params);
+
+  // Not copyable or movable.
+  simple_data_iterator(const simple_data_iterator&) = delete;
+  simple_data_iterator& operator=(const simple_data_iterator&) = delete;
+
+  const flex_list& feature_names() const override;
+  const flex_list& class_labels() const override;
+  flex_type_enum session_id_type() const override;
+  bool has_next_batch() const override;
+  batch next_batch(size_t batch_size) override;
+  void reset() override;
+
+private:
+
+  struct preprocessed_data {
+    gl_sframe chunks;
+    bool has_target = false;
+    flex_list class_labels;
+  };
+
+  static preprocessed_data preprocess_data(const parameters& params);
+
+  const preprocessed_data data_;
+  
+  gl_sframe_range range_iterator_;
+  gl_sframe_range::iterator next_row_;
+  gl_sframe_range::iterator end_of_rows_;
+  size_t sample_in_row_ = 0;
```

Comment:
So this was added for AC to have data augmentation. You should maybe look at OD for iterator stuff!

Reply:
For the iterator specifically, OD (without data augmentation or annotations) is a closer fit for what I expect DC needs. Please don't copy and paste blindly; only copy (or factor out!) what you need!

Note that we don't necessarily need to define this class in the initial commit. The entire data iterator class (even its header) is an implementation detail from the point of view of the `drawing_classifier` class